pygments/lexers/markup.py

```diff
@@ -760,3 +763,69 @@ def _handle_cssblock(self, match):
     def __init__(self, **options):
         self.handlecodeblocks = get_bool_opt(options, 'handlecodeblocks', True)
         RegexLexer.__init__(self, **options)
+
+class TexinfoLexer(RegexLexer):
+    """
+    For `Texinfo <https://gnu.org/software/texinfo>`_ source code.
+
+    .. versionadded:: 2.10
+    """
+
+    name = 'Texinfo'
+    aliases = ['texi', 'texinfo']
+    filenames = ['*.texi']
+    mimetypes = ['text/x-texi']
+
+    tokens = {
+        'base': [
+            (r'\s+', Text),
+            (r'@$', Keyword),
+            (r'@[@:{}*]', Keyword),
+            (r'(?s)``.*?''', String),
+        ],
+        'commands': [
+            (r'(@[a-z]+)(\{)',
+             bygroups(Keyword, Punctuation), 'command'),
+            (r'@[a-z]+', Keyword)
+        ],
+        'root': [
+            (r'@c\s.*?$', Comment),
+            (r'@ignore.*?@end ignore', Comment),
+            include('base'),
+            (r'(@)(node|section|subsection|subsubsection|part|chapter|centerchap|unnumbered|unnumberedsec|unnumberedsubsec|unnumberedsubsection|unnumberedsubsubsec|unnumberedsubsubsection|chapheading|majorheading|heading|subheading|subsubheading|appendix|appendixsec|appendixsection|appendixsubsec|appendixsubsection|appendixsubsubsec|appendixsubsubsection|[cfvkpt]index|printindex|syncodeindex)(\s)',
+             bygroups(Keyword, Keyword, Text.Whitespace), 'label'),
+            (r'(@include)(\s)(.*)$',
+             bygroups(Keyword, Text.Whitespace, String)),
+            (r'(?s)(@p?x?ref)(\{)([^}]+)(\})',
+             bygroups(Keyword, Punctuation, Name.Label, Punctuation)),
+            (r'(@end)(\s)([a-z]+)',
+             bygroups(Keyword, Text.Whitespace, Name.Builtin)),
+            (r'(@)(example|smallexample|menu)(\s|$)',
+             bygroups(Keyword, Keyword, Text.Whitespace), 'verbatim'),
+            (r'(@lisp)(\s|$)',
+             bygroups(Keyword, Text.Whitespace), 'lisp'),
+            include('commands')
+        ],
+        'label': [
+            include('base'),
+            (r'[^@]*?$', Name.Label, '#pop'),
+            include('commands')
+        ],
+        'command': [
+            include('base'),
+            (r'\}', Punctuation, '#pop'),
+            include('commands'),
+            (r'[^@{}]+', Name.Attribute)
```

Comment:
I believe you need to push when seeing `{`. You can have something like `@code{nested {braces}}`.